reco_utils/dataset/pandas_df_utils.py

```diff
@@ -74,3 +75,75 @@ def filter_by(df, filter_by_df, filter_by_cols):
             filter_by_df.set_index(filter_by_cols).index
         )
     ]
+
+
+def df_to_libffm(df, col_rating=DEFAULT_RATING_COL, filepath=None):
+    """Converts an input Dataframe (df) to an text file in libffm format.
+
+    Note:
+        The input dataframe is expected to represent the feature data in the following schema
+        |field-1|field-2|...|field-n|rating|
+        |feature-1-1|feature-2-1|...|feature-n-1|1|
+        |feature-1-2|feature-2-2|...|feature-n-2|0|
+        ...
+        |feature-1-i|feature-2-j|...|feature-n-k|0|
+        Where
+        1. each "field-*" occupies one column in the data, and
+        2. "feature-*-*" can be either a string or a numerical value.
+
+        The above data will be converted to the libffm format by following the convention as explained in
+        https://www.csie.ntu.edu.tw/~r01922136/slides/ffm.pdf
+
+    Args:
+        df (pd.DataFrame): input Pandas dataframe.
+        col_rating (str): rating of the data.
+        filepath (str): path to save the converted data.
+
+    Return:
+        df_libffm (pd.DataFrame): a Pandas DataFrame in libffm format.
+    """
+    df_new = df.copy()
+
+    # Encode field (rating column is not included)
+    field_names = list(df_new.drop(col_rating, axis=1).columns)
+    field_dict = {k: v+1 for v, k in enumerate(field_names)}
+
+    # Encode field-feature
+    field_feature_list = []
+    for col in field_names:
+        if df_new[col].dtype == int:
+            df_new[col] = df_new[col].astype(str)
+        if df_new[col].dtype == object:
+            field_feature_concat = ['{}-{}'.format(col, x) for x in df_new[col].values]
+            field_feature_list.extend(field_feature_concat)
+    field_feature_dict = {k: v+1 for v, k in enumerate(field_feature_list)}
+
+    def _convert(field, feature, field_index_dict, field_feature_index_dict):
+        field_index = field_index_dict[field]
+        field_feature_index = (
+            field_feature_index_dict['{}-{}'.format(field, feature)]
+            if isinstance(feature, str)
+            else None
+        )
+
+        return (
+            "{}:{}:1".format(field_index, field_feature_index)
+            if isinstance(feature, str)
+            else "{}:{}:{}".format(field_index, field_index, feature)
+        )
+
+    for col in field_names:
+        if col is not col_rating:
+            df_new[col] = df_new[col].apply(lambda x: _convert(col, x, field_dict, field_feature_dict))
+
+    # Move rating column to the first.
+    field_names.append(col_rating)
+    field_names.insert(0, field_names.pop(field_names.index(col_rating)))
+    df_new = df_new[field_names]
+
+    if filepath is not None:
+        np.savetxt(filepath, df_new.values, delimiter=' ', fmt='%s')
+
+    return df_new
```

Comment:
LIBFFM is a sparse format that is difficult to represent as a Pandas dataframe. You represent only the nonzero values, which can save a lot of space, and the representation of each data point varies in length. The closest data representation is a scipy sparse matrix, but even that would be unnecessarily cumbersome. I think we should keep it simple and write as csv or tsv file.

Reply:
Hi @anargyri I am not sure if we need to represent the zero ones in the converted data. In the [paper](https://www.csie.ntu.edu.tw/~cjlin/papers/ffm.pdf), FFM has variable number of **nfk** where **n**, **f**, **k** are the number of non-zero elements, fields, latent factors, respectively.  

Also, in Section 3.3 of the paper, the first sub-section, it says
> Note that according to the number of possible values in a categorical feature, the same number of binary features are generated and every time only one of them has the value 1. In the LIBSVM format, features with zero values are not stored. We apply the same setting to all models, so in this paper, every categorical feature is transformed to several binary ones. 

The writing to file is to benefit parsing by the deeprec algorithms such as xDeepFM - it reads text file directly.

Would love to have your advice!

Reply:
What I was saying is that you should not represent the zeros; but a pandas DF will inevitably have zeros in it.
Better sync up with @Leavingseason and also take a look at the example in https://www.csie.ntu.edu.tw/~r01922136/slides/ffm.pdf
In general, LIBFFM represents cases like that in a sparse format.

Reply:
Hmmm I am not sure if I understand... 'zeros' here refer to those in the 'field-feature' tuples? Are they needed explicitly in the libffm format? I can avoid using pandas DF and directly write a text file. The purpose of having a pandas DF is for the convenience of splitting with our repo utils. 

In the slides example, it shows a case that is not captured here, which is more-than-one features in an entry of a particular field. That is, in the movielens dataset, the genre of a movie is 'Comedy|Action|...'. In this case, I am not sure whether it is good to leave the multi-feature column encoding to another preparation function before the data is converted, or wrap the thing into this function... For example, for this movielens dataset, we may apply an encoding operation on the genres to encode them in a consistent way (example as you can see in the w&d notebook), and then it can be regarded as a new field with unique features in each column entry, which is the standard format expected by the function. 

What do you (@anargyri @loomlike @Leavingseason @yexing99) think?

Reply:
See my comments below.